[SYCL-Graph][Doc] executable graphs could not be executed at the same time (#11675)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

_executable_ by the user invoking `command_graph::finalize()` to create a
new executable instance of the graph. An executable graph cannot be converted
to a modifiable graph. After finalizing a graph in the modifiable state, it is

> valid for a user to add additional nodes into the modifialbe graph and finalize

Reviewer: Typo:

```suggestion
valid for a user to add additional nodes into the modifiable graph and finalize
```

context associated with the graph. It is valid to call this method multiple times
to create subsequent executable graphs. It is also valid to continue to add new
nodes to the modifiable graph instance after calling this function. It is valid
to finalize an empty graph instance with no recorded commands. The created
executable graph(s) from the same non-empty modifiable graph could not be executed

> simultaneously because they use the same memory address recorded in the modifiable

Reviewer 2: I think "use" here is too general. Is this UB not only an issue if one of the graphs is doing a write to the memory address? If the graphs were only reading memory would this be an issue? (I know this is contrived) 
Prefer something more specific, like
> should not be executed simultaneously if at least one executable graph contains a write operation to a memory address accessed by another executable graph

Reviewer: 1. If a modifiable graph is modified after finalization, is it considered to be the same or not?

2. Documentation for `handler::ext_oneapi_graph` says "Concurrent graph execution can be achieved by finalizing a graph in modifiable state into multiple graphs in executable state." Which, on the surface, contradicts this statement here (at least in the common case of two copies of the graph writing to the same memory location)?

Author: > 1. If a modifiable graph is modified after finalization, is it considered to be the same or not?
It depends if a same memory write is in the executions, to be accurate, I'll remove "from the same non-empty modifiable graph"

> 2. Documentation for `handler::ext_oneapi_graph` says "Concurrent graph execution can be achieved by finalizing a graph in modifiable state into multiple graphs in executable state." Which, on the surface, contradicts this statement here (at least in the common case of two copies of the graph writing to the same memory location)?

nice catch, I think the concurrent graph execution is not allowed for most cases, wait for more comments

Reviewer: > nice catch, I think the concurrent graph execution is not allowed for most cases, wait for more comments

And to clarify, the problem is only with execution, not submission? It is legal to submit two executable graphs created from the same modifiable graph are submitted into an in-order queue one after the other, right?

Author: > > nice catch, I think the concurrent graph execution is not allowed for most cases, wait for more comments
> 
> And to clarify, the problem is only with execution, not submission? It is legal to submit two executable graphs created from the same modifiable graph are submitted into an in-order queue one after the other, right?

yes, I think so.  It is legal to submit the  executable graph(s) again and again into an in-order queue, there will be no race condition for the device memory.

Reviewer 2: Something to keep in mind for this PR is that in future a graph update API could modify the arguments used by an executable graph. So we want the wording about UB to be defined in terms of the race conditions accessing the memory used by concurrent executable graphs, rather than focusing on the modifiable graph they were created from. I think the current wording achieves that though 👍 

> And to clarify, the problem is only with execution, not submission? It is legal to submit two executable graphs created from the same modifiable graph are submitted into an in-order queue one after the other, right?

That is correct, it would also be valid to submit the executable graph(s) to an out-of-order queue with a dependency between the submissions.

> Documentation for handler::ext_oneapi_graph says "Concurrent graph execution can be achieved by finalizing a graph in modifiable state into multiple graphs in executable state." Which, on the surface, contradicts this statement here (at least in the common case of two copies of the graph writing to the same memory location)?

You are correct, good spot. I'd suggest removing this sentence as part of this PR.

> execution, it is undefined behavior to do so.

Reviewer 3: This language seems a bit low level to me given that the user is not in direct control of the execution of a graph - that is ultimately decided by the SYCL runtime scheduler and backend. In the case of graphs which use buffers with write access, the scheduler will create dependencies between graph submissions anyway and prevent simultaneous execution to correctly enforce dependencies. For USM this seems to be no different than normal usage - the user is required to enforce dependencies themselves with events to ensure correct behaviour. 

It seems inappropriate to me to categorize incorrect usage of USM as undefined behaviour for graphs specifically, unless I am misunderstanding something here.

Author: agree that the buffer/access case and USM case do not have meaningful difference here.  For the USM case, the dependency is achieved by in-order queue internally or by user explicitly.

The change here is about the concurrent execution, suppose we have a recorded graph with 3 kernels A, B and C, and kernel A reads from buffer (or USM) a and writes to b,  kernel B reads from b and writes to c, kernel C reads from c and writes to d. And the dependency is correctly set between kernels, and a, b, c and d could be USMs or buffers.

Now we have two threads to execute the graph simultaneously. For a simplified timing below:
step1):  in thread 1,  A reads from a and writes to b
step2):  in thread 2,  a is filled with new data
step3):  in thread 2,  A reads from a and writes to b
step4):  in thread 1,  B reads from b with wrong data (overwritten by thread 2)

question: will the SYCL runtime scheduler block thread 2 at step3?

this PR is to let the user know the above behavior is undefined.

btw, we'd remove 'address' in the description to also cover the buffer case.

Author: > another question is: how could the two CPU threads prepare the content of 'a' before the execution, if the concurrent execution is supported?

for this question, and also the fact that there should be an input and/or an output for the graph execution, I'll refine the wording to remove the reason since it is not easy to explain for all the cases.

@Bensuo any comment? thanks.

> graph, the behavior is undefined if we do so.

Reviewer 2: Not clear who "we" is here
```suggestion
graph, it is undefined behavior to do so.
```
